### PR TITLE
Get users

### PR DESCRIPTION
**Pull request by picollomartin**
# Summary
- Add endpoint /users GET for search users with pagination
- Add test for endpoint /users GET
- Add docs for endpoint /users GET
- Add middleware for authentication required in endpoint

### CODE REVIEW

File: training-node/app/controllers/user.js [hunk 1]
```diff
@@ -88,3 +88,18 @@ exports.login = (request, response, next) => {
     next(errors.badRequest(validation.messages));
   }
 };
+
+exports.search = (request, response, next) => {
```
Comment (gonziesc) [inline training-node/app/controllers/user.js]: usually we call it findAll or getAll

File: training-node/app/controllers/user.js [hunk 2]
```diff
@@ -88,3 +88,18 @@ exports.login = (request, response, next) => {
...
+  return userServices
+    .search(request.query.offset, request.query.limit)
+    .then(users => {
+      return userServices
+        .count()
+        .then(countUsers => {
+          response.status(200);
+          response.json({ results: users, total: countUsers });
```
Comment (gonziesc) [inline training-node/app/controllers/user.js]: I believe there is no need of the `.json`, with `res.send` its ok

File: training-node/app/controllers/user.js [hunk 3]
```diff
@@ -88,3 +88,18 @@ exports.login = (request, response, next) => {
...
+        })
+        .catch(next);
```
Comment (gonziesc) [inline training-node/app/controllers/user.js]: this catch is unnecessary

File: training-node/app/controllers/user.js [hunk 4]
```diff
@@ -88,3 +88,18 @@ exports.login = (request, response, next) => {
...
+    })
+    .catch(next);
```
Comment (gonziesc) [inline training-node/app/controllers/user.js]: errors?

File: training-node/app/services/user.js [hunk 5]
```diff
@@ -18,3 +18,7 @@ exports.create = user =>
   });
 
 exports.findByEmail = email => User.findOne({ where: { email } }).catch(notifyErrorDatabase);
+
+exports.search = (offset = 0, limit = 50) => User.findAll({ offset, limit }).catch(notifyErrorDatabase);
```
Comment (gonziesc) [inline training-node/app/services/user.js]: findAll does not return the total quantity of users, right?

File: training-node/test/users.js [hunk 6]
```diff
@@ -138,44 +180,46 @@ describe('users', () => {
...
+      successUserCreate().then(res => {
+        successUserAuth().then(response => {
+          const token = response.headers[tokenManager.HEADER_NAME];
+
+          chai
+            .request(server)
+            .get('/users?offset=0&size=2')
+            .set(tokenManager.HEADER_NAME, token)
+            .then(result => {
+              dictum.chai(result);
```
Comment (gonziesc) [inline training-node/test/users.js:194]: I'll do a test for checking that the pagination is working, counting the total users in the response should be ok

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: training-node/app/middlewares/auth.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  if (token) {
+    let payload = null;
+    try {
+      payload = tokenManager.decode(token);
+    } catch (err) {
+      logger.error(`Error decoding token ${token}`, err);
+    }
+    if (payload) {
+      userServices
```
Comment (gonziesc) [inline training-node/app/middlewares/auth.js]: return

File: training-node/test/users.js (hunk 6 above)
Comment (gonziesc) [reply training-node/test/users.js:194]: awesome

File: training-node/test/users.js [hunk 8]
```diff
@@ -138,44 +180,47 @@ describe('users', () => {
...
         .catch(err => {
-          err.should.have.status(400);
-          err.response.should.be.json;
+          err.response.should.have.status(401);
           err.response.body.should.have.property('error');
           done();
         });
     });
   });
-  it('should fail auth because user does not exist', done => {
```
Comment (gonziesc) [inline training-node/test/users.js:169]: Why this test was deleted? because it cant happen?

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: training-node/test/users.js (hunk 8 above)
Comment (picollomartin) [reply training-node/test/users.js:169]: Test still exist !!
https://github.com/wolox-training/mp-express-js/blob/get-users/training-node/test/users.js#L170
Was grouped with others tests of auth

**Review by gonziesc — APPROVED** [review]

**Pull request merged**